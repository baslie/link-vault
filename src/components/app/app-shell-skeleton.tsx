export function AppShellSkeleton() {
  return (
    <div className="flex min-h-screen flex-col bg-background">
      <header className="border-b border-border bg-card/70 backdrop-blur supports-[backdrop-filter]:bg-card/60">
        <div className="container mx-auto w-full max-w-6xl px-4 py-4 sm:px-6 lg:py-5">
          <div className="flex flex-col gap-3 sm:flex-row sm:items-center sm:justify-between">
            <div className="flex items-center gap-3">
              <div className="h-6 w-28 rounded-md bg-muted animate-pulse" />
              <div className="hidden h-3 w-44 rounded-full bg-muted/80 animate-pulse md:block" />
            </div>
            <div className="flex items-center gap-2">
              <div className="hidden h-9 w-32 rounded-md bg-muted animate-pulse sm:inline-flex" />
              <div className="h-10 w-10 rounded-md bg-muted animate-pulse" />
              <div className="h-10 w-24 rounded-md bg-muted animate-pulse" />
            </div>
          </div>
          <div className="mt-4 flex flex-col gap-3 sm:flex-row sm:items-center sm:justify-between">
            <div className="h-10 w-full rounded-md bg-muted animate-pulse sm:max-w-xl" />
            <div className="flex items-center gap-3">
              <div className="h-10 w-40 rounded-md bg-muted animate-pulse" />
              <div className="h-9 w-32 rounded-md bg-muted animate-pulse sm:hidden" />
            </div>
          </div>
        </div>
      </header>
      <main className="container mx-auto flex w-full max-w-6xl flex-1 px-4 py-10 sm:px-6 lg:py-12">
        <div className="grid w-full gap-4 sm:grid-cols-2">
          <div className="h-32 rounded-xl bg-muted/80 animate-pulse" />
          <div className="h-32 rounded-xl bg-muted/80 animate-pulse" />
        </div>
      </main>
    </div>
  );
}
