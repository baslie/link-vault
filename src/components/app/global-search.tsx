import { Search } from "lucide-react";
import type { HTMLAttributes } from "react";

import { Input } from "@/components/ui/input";
import { cn } from "@/lib/utils";

type GlobalSearchProps = HTMLAttributes<HTMLDivElement>;

export function GlobalSearch({ className, ...props }: GlobalSearchProps) {
  return (
    <div className={cn("w-full sm:max-w-xl", className)} {...props}>
      <label className="sr-only" htmlFor="global-search">
        Глобальный поиск по ссылкам
      </label>
      <div className="relative">
        <Search
          aria-hidden
          className="pointer-events-none absolute left-3 top-1/2 h-4 w-4 -translate-y-1/2 text-muted-foreground"
        />
        <Input
          id="global-search"
          type="search"
          placeholder="Поиск по ссылкам (в разработке)"
          readOnly
          aria-disabled="true"
          title="Поиск появится после реализации API."
          className="cursor-not-allowed border-dashed pl-10 pr-4 text-muted-foreground"
          aria-describedby="global-search-helper"
        />
      </div>
      <p id="global-search-helper" className="mt-1 text-xs text-muted-foreground">
        Возможность искать появится после запуска API и индексов.
      </p>
    </div>
  );
}
