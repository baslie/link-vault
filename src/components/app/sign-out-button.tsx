import { signOut } from "@/app/auth/actions";
import { Button } from "@/components/ui/button";

export function SignOutButton() {
  return (
    <form action={signOut}>
      <Button type="submit" variant="outline" size="sm">
        Выйти
      </Button>
    </form>
  );
}
