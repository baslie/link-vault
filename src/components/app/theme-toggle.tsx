"use client";

import { Moon, Sun } from "lucide-react";
import { useEffect, useState } from "react";
import { useTheme } from "next-themes";

import { Button } from "@/components/ui/button";

export function ThemeToggle() {
  const { resolvedTheme, theme, setTheme } = useTheme();
  const [isMounted, setIsMounted] = useState(false);

  useEffect(() => {
    setIsMounted(true);
  }, []);

  const activeTheme = (isMounted ? resolvedTheme ?? theme : theme) ?? "light";
  const isDark = activeTheme === "dark";

  function handleToggle() {
    setTheme(isDark ? "light" : "dark");
  }

  return (
    <Button
      type="button"
      variant="outline"
      size="icon"
      onClick={handleToggle}
      aria-label="Переключить тему"
      title={isDark ? "Переключить на светлую тему" : "Переключить на тёмную тему"}
    >
      {isDark ? (
        <Moon aria-hidden className="h-[1.15rem] w-[1.15rem]" />
      ) : (
        <Sun aria-hidden className="h-[1.2rem] w-[1.2rem]" />
      )}
    </Button>
  );
}
