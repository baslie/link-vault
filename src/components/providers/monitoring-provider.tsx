"use client";

import { useEffect } from "react";

import { logMonitoringEvent } from "@/lib/monitoring/client";

export function MonitoringProvider() {
  useEffect(() => {
    function handleError(event: ErrorEvent) {
      void logMonitoringEvent({
        eventType: "client_error",
        details: {
          message: event.message,
          stack: event.error instanceof Error ? (event.error.stack ?? undefined) : undefined,
          source: event.filename ?? undefined,
          line: event.lineno ?? undefined,
          column: event.colno ?? undefined,
        },
        context: {
          userAgent: typeof navigator !== "undefined" ? navigator.userAgent : undefined,
          url: typeof window !== "undefined" ? window.location.href : undefined,
        },
      });
    }

    function handleRejection(event: PromiseRejectionEvent) {
      const reason = event.reason;
      const message = reason instanceof Error ? reason.message : String(reason);
      const stack = reason instanceof Error ? (reason.stack ?? undefined) : undefined;

      void logMonitoringEvent({
        eventType: "client_error",
        details: {
          message,
          stack,
        },
        context: {
          userAgent: typeof navigator !== "undefined" ? navigator.userAgent : undefined,
          url: typeof window !== "undefined" ? window.location.href : undefined,
        },
      });
    }

    window.addEventListener("error", handleError);
    window.addEventListener("unhandledrejection", handleRejection);

    return () => {
      window.removeEventListener("error", handleError);
      window.removeEventListener("unhandledrejection", handleRejection);
    };
  }, []);

  return null;
}
