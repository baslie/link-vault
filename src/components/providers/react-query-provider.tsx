"use client";

import { ReactQueryDevtools } from "@tanstack/react-query-devtools";
import { QueryClient, QueryClientProvider, type DefaultOptions } from "@tanstack/react-query";
import { type ReactNode, useState } from "react";

const defaultOptions: DefaultOptions = {
  queries: {
    refetchOnWindowFocus: false,
    staleTime: 1000 * 60,
  },
};

export function ReactQueryProvider({ children }: { children: ReactNode }) {
  const [client] = useState(
    () =>
      new QueryClient({
        defaultOptions,
      }),
  );

  return (
    <QueryClientProvider client={client}>
      {children}
      {process.env.NODE_ENV === "development" ? (
        <ReactQueryDevtools initialIsOpen={false} position="bottom" />
      ) : null}
    </QueryClientProvider>
  );
}
