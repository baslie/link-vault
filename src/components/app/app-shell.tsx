import Link from "next/link";
import type { ReactNode } from "react";

import { GlobalSearch } from "@/components/app/global-search";
import { SignOutButton } from "@/components/app/sign-out-button";
import { ThemeToggle } from "@/components/app/theme-toggle";
import { Button } from "@/components/ui/button";
import type { Tables } from "@/lib/supabase/types";

type Profile = Pick<Tables<"profiles">, "id" | "email" | "display_name" | "theme">;

interface AppShellProps {
  profile: Profile | null;
  children: ReactNode;
}

export function AppShell({ profile, children }: AppShellProps) {
  const displayName = profile?.display_name?.trim() || profile?.email || "Новый пользователь";
  const email = profile?.email;

  return (
    <div className="flex min-h-screen flex-col bg-background">
      <header className="border-b border-border bg-card/70 backdrop-blur supports-[backdrop-filter]:bg-card/60">
        <div className="container mx-auto flex w-full max-w-6xl flex-col gap-4 px-4 py-4 sm:px-6 lg:py-5">
          <div className="flex flex-col gap-3 sm:flex-row sm:items-center sm:justify-between">
            <div className="flex items-center gap-3">
              <Link href="/app" className="text-lg font-semibold tracking-tight">
                Link Vault
              </Link>
              <span className="hidden text-sm text-muted-foreground md:inline">
                Минималистичный менеджер ссылок
              </span>
            </div>
            <div className="flex items-center gap-2 sm:justify-end">
              <Button type="button" size="sm" className="hidden sm:inline-flex" disabled>
                Добавить ссылку
              </Button>
              <ThemeToggle />
              <SignOutButton />
            </div>
          </div>
          <div className="flex flex-col gap-3 sm:flex-row sm:items-center sm:justify-between">
            <GlobalSearch className="sm:flex-1" />
            <div className="flex w-full items-center justify-between gap-3 sm:w-auto sm:justify-end">
              <div className="text-sm text-muted-foreground sm:text-right">
                <p className="font-medium text-foreground">{displayName}</p>
                {email ? <p className="text-xs">{email}</p> : <p className="text-xs">Профиль загружается...</p>}
              </div>
              <Button type="button" size="sm" className="sm:hidden" disabled>
                Добавить ссылку
              </Button>
            </div>
          </div>
        </div>
      </header>
      <main className="container mx-auto flex w-full max-w-6xl flex-1 px-4 py-10 sm:px-6 lg:py-12">{children}</main>
    </div>
  );
}
