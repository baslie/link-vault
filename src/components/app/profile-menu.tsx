"use client";

import { useEffect, useRef, useState } from "react";
import Link from "next/link";
import { LogOut, Settings, ShieldCheck, Upload } from "lucide-react";

import { ThemeToggle } from "@/components/app/theme-toggle";
import { SignOutButton } from "@/components/app/sign-out-button";
import type { ProfileRecord } from "@/lib/profile/schema";

interface ProfileMenuProps {
  profile: ProfileRecord | null;
  isAdmin?: boolean;
}

function buildInitials(name?: string | null, email?: string | null) {
  const source = name?.trim() || email?.trim() || "";
  if (!source) {
    return "?";
  }

  const [first] = source.split(" ");
  return first.slice(0, 2).toUpperCase();
}

export function ProfileMenu({ profile, isAdmin = false }: ProfileMenuProps) {
  const [isOpen, setIsOpen] = useState(false);
  const containerRef = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    function handleClick(event: MouseEvent) {
      if (!containerRef.current?.contains(event.target as Node)) {
        setIsOpen(false);
      }
    }

    function handleKey(event: KeyboardEvent) {
      if (event.key === "Escape") {
        setIsOpen(false);
      }
    }

    document.addEventListener("click", handleClick);
    document.addEventListener("keydown", handleKey);
    return () => {
      document.removeEventListener("click", handleClick);
      document.removeEventListener("keydown", handleKey);
    };
  }, []);

  const initials = buildInitials(profile?.display_name, profile?.email);

  return (
    <div className="relative" ref={containerRef}>
      <button
        type="button"
        className="flex h-10 w-10 items-center justify-center rounded-full bg-primary/10 text-sm font-medium text-primary shadow-sm transition hover:bg-primary/20"
        onClick={() => setIsOpen((prev) => !prev)}
        aria-haspopup="menu"
        aria-expanded={isOpen}
      >
        {initials}
      </button>
      {isOpen ? (
        <div className="absolute right-0 z-50 mt-3 w-56 rounded-lg border border-border bg-popover shadow-lg">
          <div className="flex flex-col gap-1 p-3 text-sm">
            <div className="mb-2 border-b border-border pb-2">
              <p className="font-medium text-foreground">{profile?.display_name ?? "Профиль"}</p>
              <p className="text-xs text-muted-foreground">{profile?.email ?? "Email не указан"}</p>
            </div>
            <Link
              href="/app/settings"
              className="flex items-center gap-2 rounded-md px-2 py-2 text-sm text-foreground transition hover:bg-muted"
              onClick={() => setIsOpen(false)}
            >
              <Settings className="h-4 w-4" aria-hidden />
              Настройки профиля
            </Link>
            {isAdmin ? (
              <Link
                href="/admin"
                className="flex items-center gap-2 rounded-md px-2 py-2 text-sm text-foreground transition hover:bg-muted"
                onClick={() => setIsOpen(false)}
              >
                <ShieldCheck className="h-4 w-4" aria-hidden />
                Админ-панель
              </Link>
            ) : null}
            <Link
              href="/app/import"
              className="flex items-center gap-2 rounded-md px-2 py-2 text-sm text-foreground transition hover:bg-muted"
              onClick={() => setIsOpen(false)}
            >
              <Upload className="h-4 w-4" aria-hidden />
              Импорт CSV
            </Link>
            <Link
              href="/app/export"
              className="flex items-center gap-2 rounded-md px-2 py-2 text-sm text-foreground transition hover:bg-muted"
              onClick={() => setIsOpen(false)}
            >
              <Upload className="h-4 w-4 rotate-180" aria-hidden />
              Экспорт CSV
            </Link>
            <div className="mt-1 border-t border-border pt-2">
              <div className="flex items-center justify-between px-2 py-2 text-sm text-foreground">
                <span>Тема</span>
                <ThemeToggle />
              </div>
              <SignOutButton
                variant="ghost"
                className="mt-2 flex w-full items-center gap-2 px-2 py-2 text-sm text-destructive hover:bg-destructive/10"
                icon={<LogOut className="h-4 w-4" aria-hidden />}
              >
                Выйти
              </SignOutButton>
            </div>
          </div>
        </div>
      ) : null}
    </div>
  );
}
