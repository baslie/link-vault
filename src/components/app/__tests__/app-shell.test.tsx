vi.mock("@/components/app/sign-out-button", () => ({
  SignOutButton: () => <button type="button">Выйти</button>,
}));

import { fireEvent, render, screen } from "@testing-library/react";

import { AppShell } from "@/components/app/app-shell";
import { ThemeProvider } from "@/components/providers/theme-provider";

const profile = {
  id: "123",
  email: "user@example.com",
  display_name: "Анна", // eslint-disable-line camelcase -- соответствие полю таблицы
  theme: "system" as const,
};

describe("AppShell", () => {
  function renderShell() {
    return render(
      <ThemeProvider>
        <AppShell profile={profile} isAdmin>
          <div>Контент</div>
        </AppShell>
      </ThemeProvider>,
    );
  }

  it("отображает бренд, профиль и поисковую заглушку", () => {
    renderShell();

    expect(screen.getByRole("link", { name: "Link Vault" })).toBeVisible();
    expect(screen.getByText("Минималистичный менеджер ссылок")).toBeVisible();
    expect(screen.getByText(profile.email)).toBeVisible();
    expect(screen.getByRole("searchbox", { name: /глобальный поиск/i })).toHaveAttribute(
      "aria-disabled",
      "true",
    );
    fireEvent.click(screen.getByRole("button", { name: /ан/i }));
    expect(screen.getByRole("link", { name: "Админ-панель" })).toBeVisible();
  });

  it("соответствует снепшоту", () => {
    const { asFragment } = renderShell();

    expect(asFragment()).toMatchSnapshot();
  });
});
