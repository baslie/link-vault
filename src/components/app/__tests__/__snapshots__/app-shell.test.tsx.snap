// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppShell > соответствует снепшоту 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="flex min-h-screen flex-col bg-background"
  >
    <header
      class="border-b border-border bg-card/70 backdrop-blur supports-[backdrop-filter]:bg-card/60"
    >
      <div
        class="container mx-auto flex w-full max-w-6xl flex-col gap-4 px-4 py-4 sm:px-6 lg:py-5"
      >
        <div
          class="flex flex-col gap-3 sm:flex-row sm:items-center sm:justify-between"
        >
          <div
            class="flex items-center gap-3"
          >
            <a
              class="text-lg font-semibold tracking-tight"
              href="/app"
            >
              Link Vault
            </a>
            <span
              class="hidden text-sm text-muted-foreground md:inline"
            >
              Минималистичный менеджер ссылок
            </span>
          </div>
          <div
            class="flex items-center gap-2 sm:justify-end"
          >
            <button
              class="items-center justify-center whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 ring-offset-background bg-primary text-primary-foreground hover:bg-primary/90 h-9 rounded-md px-3 hidden sm:inline-flex"
              disabled=""
              type="button"
            >
              Добавить ссылку
            </button>
            <button
              aria-label="Переключить тему"
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 ring-offset-background border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10"
              title="Переключить на тёмную тему"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-sun h-[1.2rem] w-[1.2rem]"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="4"
                />
                <path
                  d="M12 2v2"
                />
                <path
                  d="M12 20v2"
                />
                <path
                  d="m4.93 4.93 1.41 1.41"
                />
                <path
                  d="m17.66 17.66 1.41 1.41"
                />
                <path
                  d="M2 12h2"
                />
                <path
                  d="M20 12h2"
                />
                <path
                  d="m6.34 17.66-1.41 1.41"
                />
                <path
                  d="m19.07 4.93-1.41 1.41"
                />
              </svg>
            </button>
            <button
              type="button"
            >
              Выйти
            </button>
          </div>
        </div>
        <div
          class="flex flex-col gap-3 sm:flex-row sm:items-center sm:justify-between"
        >
          <div
            class="w-full sm:max-w-xl sm:flex-1"
          >
            <label
              class="sr-only"
              for="global-search"
            >
              Глобальный поиск по ссылкам
            </label>
            <div
              class="relative"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-search pointer-events-none absolute left-3 top-1/2 h-4 w-4 -translate-y-1/2 text-muted-foreground"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="11"
                  cy="11"
                  r="8"
                />
                <path
                  d="m21 21-4.3-4.3"
                />
              </svg>
              <input
                aria-describedby="global-search-helper"
                aria-disabled="true"
                class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 cursor-not-allowed border-dashed pl-10 pr-4 text-muted-foreground"
                id="global-search"
                placeholder="Поиск по ссылкам (в разработке)"
                readonly=""
                title="Поиск появится после реализации API."
                type="search"
              />
            </div>
            <p
              class="mt-1 text-xs text-muted-foreground"
              id="global-search-helper"
            >
              Возможность искать появится после запуска API и индексов.
            </p>
          </div>
          <div
            class="flex w-full items-center justify-between gap-3 sm:w-auto sm:justify-end"
          >
            <div
              class="text-sm text-muted-foreground sm:text-right"
            >
              <p
                class="font-medium text-foreground"
              >
                Анна
              </p>
              <p
                class="text-xs"
              >
                user@example.com
              </p>
            </div>
            <button
              class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 ring-offset-background bg-primary text-primary-foreground hover:bg-primary/90 h-9 rounded-md px-3 sm:hidden"
              disabled=""
              type="button"
            >
              Добавить ссылку
            </button>
          </div>
        </div>
      </div>
    </header>
    <main
      class="container mx-auto flex w-full max-w-6xl flex-1 px-4 py-10 sm:px-6 lg:py-12"
    >
      <div>
        Контент
      </div>
    </main>
  </div>
</DocumentFragment>
`;
